### DISCUSSION TITLE
Captcha in forms

### DISCUSSION THREAD

**Discussion by timotius02** [remix-run/remix#2716] [Q&A]
Does anyone have any examples of using Captcha in Remix Forms? Was trying to work with the react-google-recaptcha component but I was having issues getting my site key to render correctly.

**Comment by datenmassiv** [comment]
Hi @timotius02 ,
not sure what the actual problem is you ran into as your problem description is a bit vague.
But maybe check on how Browser Environment variables are exposed to the browser if you use environment variables for the google site key:
https://remix.run/docs/en/v1/guides/envvars#browser-environment-variables

**Reply by timotius02** [reply]
I was just asking if there were any example projects out there using captchas with Remix. I remember that their suggestion there didn't work for me. I ended loading my keys up in a loader and passing to the component via useLoaderData but that felt really bad since  the rest of the page was basically static and didn't need that overhead.

**Comment by manV** [comment]
I recently integrated recaptcha enterprise checkbox on one of my forms. Here is what I did.

in root.tsx file under `<head>` tag
```html
<script src={`https://www.google.com/recaptcha/enterprise.js?render=explicit`}></script>
```

and in your form component

```js
const MyForm = () => {
  const grecaptchaRef = useRef();
  useEffect(() => {
    (window as any)?.grecaptcha?.enterprise?.ready?.(() => {
      (window as any)?.grecaptcha?.enterprise?.render?.(grecaptchaRef.current, {
        sitekey: (window as any)?.ENV?.RECAPTCHA_SITE_KEY,
        action: 'submit',
      });
    });
  }, []);


  return (
    <Form action="post" replace>
      {... other form fields}
      <div ref={gcaptchaRef}/>
      <button type="submit">submit</button>
    </Form>
  );
}
```

and then on submitting the form, you should have `g-recaptcha-response` as a part of formdata in your actions. you can check the token for validating with recaptcha assessment api.

I followed [this](https://remix.run/docs/en/v1/guides/envvars#browser-environment-variables) to add client side environment variables.

hope this helps.
